src/realm/group_writer.cpp

```diff
@@ -687,8 +687,12 @@ std::pair<size_t, size_t> GroupWriter::extend_free_space(size_t requested_size)
     // extended the file size. It can also happen as part of initial file expansion
     // during attach_file().
     size_t logical_file_size = to_size_t(m_group.m_top.get(2) / 2);
-    size_t extend_size = requested_size;
-    size_t new_file_size = logical_file_size + extend_size;
+    size_t new_file_size = logical_file_size;
+    if (REALM_UNLIKELY(int_add_with_overflow_detect(new_file_size, requested_size))) {
+        throw MaxAddressBreached("GroupWriter cannot extend free space: " + util::to_string(logical_file_size) + " + "
+                                 + util::to_string(requested_size));
+    }
+
     if (!alloc.matches_section_boundary(new_file_size)) {
         new_file_size = alloc.get_upper_section_boundary(new_file_size);
     }
```

Comment:
@finnschiermer It seemed to me that the upper section boundary was returning a `size_t` and so should never overflow so we wouldn't have to check there. However, your point lead me to test the max edge case here when the `new_file_size` grows to the maximum upper section boundary. It turns out that the maximum upper boundary is not calculated correctly and the existing code could actually end up _decreasing_ `new_file_size`. That could be the root of some strange failures we see including cases where we observe the logical file size to be greater than the actual file size. I've committed a failing unit test to this branch, let me know if you have any ideas to fix it otherwise I'll look for a fix tomorrow (we might have to special handle the calculation of the last section or else limit allocations to the second to last section).